src/RWMutex.hpp

```diff
@@ -109,6 +121,86 @@ class _LIBUNWIND_HIDDEN RWMutex {
 
 #endif
 
+/// Simple lock-free reader-writer lock with writer preference.
+/// It's supposed to be async-signal-safe.
+class LockFreeRWMutex {
+public:
+  LockFreeRWMutex() {
+    atomic_init(&state, 0);
+    atomic_init(&waiting_writers, 0);
+  }
+
+  bool lock_shared() {
+    while (true) {
+      int current_state = atomic_load(&state);
+      if (current_state < 0 || current_state >= INT_MAX ||
+          atomic_load(&waiting_writers) > 0) {
+        sched_yield();
+        continue;
+      }
+
+      if (atomic_compare_exchange_weak(&state, &current_state,
+                                       current_state + 1)) {
+        return true;
+      }
+      sched_yield();
+    }
+  }
+
+  bool unlock_shared() {
+    int current_state = atomic_fetch_sub(&state, 1);
+    if (current_state <= 0) {
+      abort();
+    }
+    return true;
+  }
+
+  bool lock() {
+    while (true) {
+      int current_waiting_writers = atomic_load(&waiting_writers);
+      if (current_waiting_writers == INT_MAX) {
+        sched_yield();
+        continue;
+      }
+
+      if (atomic_compare_exchange_weak(&waiting_writers,
+                                       &current_waiting_writers,
+                                       current_waiting_writers + 1)) {
+        break;
+      }
+      sched_yield();
+    }
+
+    while (true) {
+      int expected = 0;
+      if (atomic_compare_exchange_weak(&state, &expected, -1)) {
+        atomic_fetch_sub(&waiting_writers, 1);
```

Comment:
Nitpick (EDIT: meh, merging without it):
```suggestion
        int current_waiting_writers = atomic_fetch_sub(&waiting_writers, 1);
        if (current_waiting_writers <= 0) {
          abort();
        }
```